examples/defaulter-gen/generators/defaulter.go

```diff
@@ -0,0 +1,710 @@
+}
+
+func (g *genDefaulter) Namers(c *generator.Context) namer.NameSystems {
+	// Have the raw namer for this file track what it imports.
+	return namer.NameSystems{
+		"raw": namer.NewRawNamer(g.targetPackage, g.imports),
+	}
+}
+
+func (g *genDefaulter) isOtherPackage(pkg string) bool {
+	if pkg == g.targetPackage {
+		return false
+	}
+	if strings.HasSuffix(pkg, `"`+g.targetPackage+`"`) {
+		return false
+	}
+	return true
+}
+
+func (g *genDefaulter) Filter(c *generator.Context, t *types.Type) bool {
+	defaults, ok := g.newDefaulters[t]
+	if !ok || defaults.object == nil {
+		return false
+	}
+	g.typesForInit = append(g.typesForInit, t)
+	return true
+}
+
+func (g *genDefaulter) Imports(c *generator.Context) (imports []string) {
+	var importLines []string
+	for _, singleImport := range g.imports.ImportLines() {
+		if g.isOtherPackage(singleImport) {
+			importLines = append(importLines, singleImport)
+		}
+	}
+	return importLines
+}
+
+func (g *genDefaulter) Init(c *generator.Context, w io.Writer) error {
+	sw := generator.NewSnippetWriter(w, c, "$", "$")
+
+	scheme := c.Universe.Type(types.Name{Package: runtimePackagePath, Name: "Scheme"})
+	schemePtr := &types.Type{
+		Kind: types.Pointer,
+		Elem: scheme,
+	}
+	sw.Do("// RegisterDefaults adds defaulters functions to the given scheme.\n", nil)
+	sw.Do("// Public to allow building arbitrary schemes.\n", nil)
+	sw.Do("// All generated defaulters are covering - they call all nested defaulters.\n", nil)
+	sw.Do("func RegisterDefaults(scheme $.|raw$) error {\n", schemePtr)
+	for _, t := range g.typesForInit {
+		args := defaultingArgsFromType(t)
+		sw.Do("scheme.AddTypeDefaultingFunc(&$.inType|raw${}, func(obj interface{}) { $.inType|objectdefaultfn$(obj.(*$.inType|raw$)) })\n", args)
+	}
+	sw.Do("return nil\n", nil)
+	sw.Do("}\n\n", nil)
+	return sw.Error()
+}
+
+func (g *genDefaulter) GenerateType(c *generator.Context, t *types.Type, w io.Writer) error {
+	if _, ok := g.newDefaulters[t]; !ok {
+		return nil
+	}
+
+	glog.V(5).Infof("generating for type %v", t)
+
+	callTree := buildCallTreeForType(t, 0, g.existingDefaulters, g.newDefaulters)
+	if callTree == nil {
+		glog.V(5).Infof("  no defaulters defined")
+		return nil
+	}
+	i := 0
+	callTree.VisitInOrder(func(ancestors []*callNode, current *callNode) {
+		if len(current.call) == 0 {
+			return
+		}
+		path := callPath(append(ancestors, current))
+		glog.V(5).Infof("  %d: %s", i, path)
+		i++
+	})
+
+	sw := generator.NewSnippetWriter(w, c, "$", "$")
+	g.generateDefaulter(t, callTree, sw)
+	return sw.Error()
+}
+
+func defaultingArgsFromType(inType *types.Type) generator.Args {
+	return generator.Args{
+		"inType": inType,
+	}
+}
+
+func (g *genDefaulter) generateDefaulter(inType *types.Type, callTree *callNode, sw *generator.SnippetWriter) {
+	sw.Do("func $.inType|objectdefaultfn$(in *$.inType|raw$) {\n", defaultingArgsFromType(inType))
+	callTree.WriteMethod("in", 0, nil, sw)
+	sw.Do("}\n\n", nil)
+}
+
+type callNode struct {
+	field string
```

Comment:
Can you please add comments to the fields (it's not clear to me what values they are storing).